<a href="https://colab.research.google.com/github/njaincode/python_for_data_science/blob/main/Describing_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Describing data with summary statistics
---

This worksheet is tied to the Pandas Getting Started Tutorials, picking out particular tutorials to link them into a theme here.

We will focus on describing data.  This is the least risky in terms of bias and inaccurate conclusions as it should focus just on what data is presented to us.

Each exercise will ask you to work through on tutorial on the Getting Started page, to try the code from the tutorial here and to try a second, similar action.

---

The practice data from the tutorials comes from a dataset on Titanic passengers.


### Exercise 1 - open the Titanic dataset
---

The Titanic dataset is stored at this URL:
https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Read the dataset into a pandas dataframe that you will call **titanic**.

**Test output**:  
The shape of the dataframe will be (891, 12)

In [ ]:
import pandas as pd

def titanic_shape():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)
  print(f'The shape of the dataframe is {titanic.shape}')

titanic_shape()

The shape of the dataframe is (891, 12)


### Exercise 2 - get summary information about the dataframe
---

Read through the tutorials:  
[What kind of data does pandas handle?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#)  
[How do I read and write tabular data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html)

Use panda functions to display the following:
1.  A technical summary of the data (info())
2.  A description of the numerical data (describe())
3.  Display the Series 'Age'

**Test output**:   
1.  The info should show that there are only 204 values in the Cabin series, out of 891 records.  
2.  The description should show 7 columns and a mean age of 29.699118
3.  The Age series should have values of type float64 and Length 891

In [ ]:
import pandas as pd

def titanic_summary():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)

  # Setup display
  pd.options.display.max_columns= 12

  print(f'titanic_info \n {titanic.info()}')

  # Description of numerical data, dtype=object do not show up
  print(f'titanic_describe \n {titanic.describe()}')
  
  print(f'titanic[age] \n {titanic["Age"]}')


titanic_summary()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
titanic_info 
 None
titanic_describe 
        PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.52300

### Exercise 3 - aggregating statistics
---

Read through the tutorial:  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

Use panda functions to display the following summary statistics from the titanic dataset:  

1.  The average (mean) age of passengers  
2.  The median age and fare  
3.  The mean fare
4.  The modal fare and gender

**Test output**:   
29.699118, Age 28.0000 Fare 14.4542, 32.2042079685746, Fare 8.05 Sex male 


In [ ]:
import pandas as pd

def titanic_stats():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)

  # Setup display
  pd.options.display.max_columns= 12

  # The average (mean) age of passengers
  df_age = titanic["Age"]
  # Observe the diff b/w mean() and mean
  df_age_avg = df_age.mean()
  df_age_median = df_age.median()

  print(f'titanic average/mean age is {df_age_avg}')
  print(f'titanic median age is {df_age_median}')

  # The median age and fare
  col_list = ["Age", "Fare"]
  df_age_fare = titanic[col_list]
  df_age_fare_median = df_age_fare.median()
  
  print(f'titanic median age X fare is \n {df_age_fare_median}')

  # The mean fare
  df_fare_mean = df_age_fare["Fare"].mean()
  print(f'titanic mean fare is {df_fare_mean}')

  # The modal fare and gender
  df_fare_gender = titanic[["Fare", "Sex"]]
  df_fare_gender_model = df_fare_gender.groupby("Sex").median()
  print(f'titanic model fare X gender is \n {df_fare_gender_model}')


titanic_stats()

titanic average/mean age is 29.69911764705882
titanic median age is 28.0
titanic median age X fare is 
 Age     28.0000
Fare    14.4542
dtype: float64
titanic mean fare is 32.2042079685746
titanic model fare X gender is 
         Fare
Sex         
female  23.0
male    10.5


### Exercise 4 - displaying other statistics
---

Take a look at the list of methods available for giving summary statistics [here](https://pandas.pydata.org/docs/user_guide/basics.html#basics-stats) 

Use panda functions, and your existing knowledge, to display the following summary statistics from the titanic dataset:

1.  The total number of passengers on the titanic
2.  The age of the youngest passenger
3.  The most expensive ticket price
4.  The range of ticket prices
5.  The number of passenges with cabins
6.  The code for the port where the highest number of passengers embarked
7.  The most populous gender
8.  The standard deviation for age and fare

**Tests**:  
891, 0.42, 512.3292, 512.3292, 204, S, male, Age 14.526497 Fare 49.693429

In [17]:
import pandas as pd

def titanic_other_stats():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)

  # Setup display
  pd.options.display.max_columns= 12

  #The total number of passengers on the titanic
  total_num_passengers = len(titanic.axes[0])
  print(f'total number of passengers on the titanic = {total_num_passengers}')

  #The age of the youngest passenger
  min_age = titanic['Age'].min()
  print(f'The age of the youngest passenger = {min_age}')

  #The most expensive ticket price
  max_price = titanic['Fare'].max()
  print(f'The most expensive ticket price {max_price}')

  #The range of ticket prices
  # FIXME

  #The number of passenges with cabins
  # Key - Cabin, its cells have text. Count number of non-empty cells
  df_cabin = titanic['Cabin']
  print(df_cabin.count())

  #The code for the port where the highest number of passengers embarked
  # Key - 'Embarked', text. Get all unique values and then count?
  df_port = titanic['Embarked']
  df_value_count = df_port.value_counts()
  #FIXME How to print S?? 
  print(f'Number of ports {df_port.nunique()}')
  print(f'Value counts {type(df_value_count)}')
  print(f'Value counts\n {df_value_count}')
  print(f'Value counts {df_value_count.iloc[0]}')

  #The most populous gender
  most_populous_gender = titanic['Sex'].mode()
  print(f'The most populous gender {most_populous_gender}')
  #FIXME How to print male?

  #The standard deviation for age and fare
  print(f'The standard deviation for age {titanic["Age"].std()}')
  print(f'The standard deviation for age {titanic["Fare"].std()}')


titanic_other_stats()

total number of passengers on the titanic = 891
The age of the youngest passenger = 0.42
The most expensive ticket price 512.3292
204
Number of ports 3
Value counts <class 'pandas.core.series.Series'>
Value counts
 S    644
C    168
Q     77
Name: Embarked, dtype: int64
Value counts 644
The most populous gender 0    male
dtype: object
The standard deviation for age 14.526497332334044
The standard deviation for age 49.693428597180905


### Exercise 5 - aggregating statistics grouped by category
---

Refer again to the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)   
looking particularly at the section on Aggregating statistics grouped by category.

1.  What is the mean age for male versus female Titanic passengers?
2.  What is the mean ticket fare price for each of the sex and cabin class combinations?
3.  What is the mean ticket fare price for passengers who embarked at each port?
4.  Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?

**Test output**:  
1.  female 27.915709 male 30.726645
2.  
```
female  1         106.125798
            2          21.970121
            3          16.118810
male    1          67.226127
            2          19.741782
            3          12.661633
```
3.  
```
C    59.954144
Q    13.276030
S    27.079812
```
4. 

```
Survived  Pclass

0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
```






In [10]:
import pandas as pd

def titanic_agg_stats():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)

  # Setup display
  pd.options.display.max_columns= 12

  #What is the mean age for male versus female Titanic passengers?
  # Sex is not numerical data
  df_age_sex = titanic[["Age", "Sex"]].groupby("Sex").mean()
  print(f'titanic average/mean age of male vs female is \n {df_age_sex}')
  
  #What is the mean ticket fare price for each of the sex and cabin class combinations?
  # First group and then select column
  df_fare_class_sex = titanic.groupby(["Sex", "Pclass"])["Fare"].mean()
  print(f'titanic mean fare for each class for male vs female is \n {df_fare_class_sex}')

  #What is the mean ticket fare price for passengers who embarked at each port?
  df_fare_port = titanic.groupby(["Embarked"])["Fare"].mean() 
  print(f'titanic mean fare for each port is \n {df_fare_port}')

  #Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?
  df_survive_pclass = titanic.groupby(["Survived"])["Pclass"].value_counts()
  print(f'titanic highest num of survivors per class \n {df_survive_pclass}')

titanic_agg_stats()

titanic average/mean age of male vs female is 
               Age
Sex              
female  27.915709
male    30.726645
titanic mean fare for each class for male vs female is 
 Sex     Pclass
female  1         106.125798
        2          21.970121
        3          16.118810
male    1          67.226127
        2          19.741782
        3          12.661633
Name: Fare, dtype: float64
titanic mean fare for each port is 
 Embarked
C    59.954144
Q    13.276030
S    27.079812
Name: Fare, dtype: float64
titanic highest num of survivors per class 
 Survived  Pclass
0         3         372
          2          97
          1          80
1         1         136
          3         119
          2          87
Name: Pclass, dtype: int64


### Exercise 6 - an aggregation of different statistics
---

Use the function titanic.agg() as shown in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

1.  Display

```
     {
         "Age": ["min", "max", "median", "skew"],
         "Fare": ["min", "max", "median", "mean"]
     }
```
2.  Display:  
min, max and mean for Age  
min, max and standard deviation for Fare  
count for Cabin

**Test output**:   
1.  	
```
                  Age	      Fare  
max	   80.000000	512.329200  
mean	  NaN	      32.204208  
median	28.000000	14.454200  
min	   0.420000	 0.000000  
skew	  0.389108	 NaN
```

2.   
```
	        Age	    Fare	   Cabin
count  NaN	    NaN	    204.0
max	80.000000  512.329200 NaN
mean   29.699118  NaN        NaN
min	0.420000   0.000000   NaN
std	NaN        49.693429  NaN
```




In [15]:
import pandas as pd

def titanic_agg_different_stats():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)

  # Setup display
  pd.options.display.max_columns= 12

  # Create a list of stats required
  display_stat1 = ["max", "mean", "median", "min", "skew"]
  display_stat2 = ["count", "max", "mean", "min", "std"]

  # Stats of Age and Fare
  df_age_fare = titanic.agg({"Age":display_stat1, "Fare":display_stat1})  
  print(f'titanic Age X Fare \n {df_age_fare}')

  stat2_dict = {"Age":display_stat2, "Fare":display_stat2, "Cabin":display_stat2}
  print(f'\ntitanic AgeXFareXCabin \n {titanic.agg(stat2_dict)}')
  
titanic_agg_different_stats()


titanic Age X Fare 
               Age        Fare
max     80.000000  512.329200
mean    29.699118   32.204208
median  28.000000   14.454200
min      0.420000    0.000000
skew     0.389108    4.787317

titanic AgeXFareXCabin 
               Age        Fare  Cabin
count  714.000000  891.000000  204.0
max     80.000000  512.329200    NaN
mean    29.699118   32.204208    NaN
min      0.420000    0.000000    NaN
std     14.526497   49.693429    NaN


### Exercise 7 - count by category
---

Read the section Count number of records by category in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)

1. Display the number of passengers of each gender who had a ticket
2. Display the number of passengers who embarked at each port and had a ticket
3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic (*Hint:  try getting the PassengerIds with a count for survived or not.  Store this value in a new variable, which will contain a list/array.  The second item in this list will be the number who survived.  You can use this number and the count of PassengerIds to calculate the percentage*)

**Test output**:  
1.  female 314, male 577
2.  C 168, Q 77, S 644
3.  38.38383838383838



In [46]:
import pandas as pd

def titanic_category_count():
  # Read csv
  csv_url = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv"
  titanic = pd.read_csv(csv_url)

  # Setup display
  pd.options.display.max_columns= 12

  # Create a list of stats required
  display_stat1 = ["max", "mean", "median", "min", "skew"]

  #1. Display the number of passengers of each gender who had a ticket
  # Note value_counts() output. It counts for each value of Ticket
  num_gender_with_ticket = titanic.groupby("Sex")["Ticket"].count()  
  print(f'Number of passengers of each gender who had a ticket \n {num_gender_with_ticket}')

  #2. Display the number of passengers who embarked at each port and had a ticket
  num_passenger_port_with_ticket = titanic.groupby("Embarked")["Ticket"].count()  
  print(f'\n Number of passengers with tickets embarked on each port \n {num_passenger_port_with_ticket}')

  #3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic
  
  # A series with survived and not survived
  ds_survived = titanic["Survived"].value_counts()
  print(f'\n Survived dataseries\n {ds_survived}')
  # Index 1 - survived
  total_num_survived = ds_survived[1]
  total_passengers = ds_survived.sum()
  survive_percent = total_num_survived*100/total_passengers
  print(f'\n Percentage of passengers survived = {survive_percent}')
   
  
titanic_category_count()


Number of passengers of each gender who had a ticket 
 Sex
female    314
male      577
Name: Ticket, dtype: int64

 Number of passengers with tickets embarked on each port 
 Embarked
C    168
Q     77
S    644
Name: Ticket, dtype: int64

 Survived dataseries 
 0    549
1    342
Name: Survived, dtype: int64

 Percentage of passengers survived = 38.38383838383838


### Exercise 8 - summary happiness statistics
---

Open the data set here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true

It contains data on people's perception of happiness levels in a number of countries across the world.

1.  Display the number of records in the set  
2.  Display the description of the numerical data  
3.  Display the highest GDP and life expectancy  
4.  Display the mean, max and min for Freedom,  mean, max, min and skew for Generosity and mean, min, max and std for GDP  

**Test output**:  
1.  156
2.  Table showing count, mean, std, min, 25%, 50%, 75%, max for 8 columns
3.  GDP 0.905147, life expectancy 0.725244  
4.  


```
	   Freedom to make life choices	Generosity	GDP per capita
max	 0.631000	                   0.566000	  1.684000
mean	0.392571	                   0.184846	  0.905147
min	 0.000000	                   0.000000  	0.000000
skew	NaN	                        0.745942	  NaN
std	 NaN	                        NaN	       0.398389
```




In [54]:
import pandas as pd

def happiness_stat():
  # Read excel
  excel_url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true"
  happiness_data = pd.read_excel(excel_url)

  # Setup display
  pd.options.display.max_columns= 12

  # 1. Display the number of records in the set
  print(f'Number of records \n {len(happiness_data)}')

  # 2. Display the description of the numerical data
  print(f'Happiness data description \n {happiness_data.describe()}')

  # 3. Display the highest GDP and life expectancy
  # FIXME Does not match expected results?
  print(f'Highest GDP and life expectancy \n {happiness_data[["GDP per capita", "Healthy life expectancy"]].max()}')

  # 4. Display the mean, max and min for Freedom, mean, max, min and skew for Generosity and mean, min, max and std for GDP
  # FIXME Test output is different from exercise description
  stat_freedom = ["mean", "max", "min"]
  stat_gen = [ "mean", "max", "min", "skew"]
  stat_gdp = ["mean", "min", "max", "std"]

  # skew & std will be NaN for freedom as it is not calculated because it is not selected in stat_freedom
  stat_dict = {"Freedom to make life choices":stat_freedom, "Generosity":stat_gen, "GDP per capita":stat_gdp}
  print(f'Freedom to make life choices \n {happiness_data.agg(stat_dict)}')

happiness_stat()

Number of records 
 156
Highest GDP and life expectancy 
 GDP per capita             1.684
Healthy life expectancy    1.141
dtype: float64
Freedom to make life choices 
       Freedom to make life choices  Generosity  GDP per capita
max                       0.631000    0.566000        1.684000
mean                      0.392571    0.184846        0.905147
min                       0.000000    0.000000        0.000000
skew                           NaN    0.745942             NaN
std                            NaN         NaN        0.398389


### Exercise 9 - migration data
---

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true  Open the sheet named *Country Migration*

1.  Describe the dataset  
2.  Show summary information
3.  Display the mean net per 10K migration in each of the years 2015 to 2019
4.  Display the mean, max and min migration for the year 2019 for each of the regions (*base_country_wb_region*)
5.  Display the median net migration for the years 2015 and 2019 for the base countries by income level
6.  Display the number of target countries in each income level
7.  Display the mean net migration for all five years, for each income level

**Test output**:  
1  count, mean, std, min, 25%, 50%, 75%, max for 9 columns

2  shows 16 columns with non-null count of 4148 in each column

3  
```
net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64
```

4  
```
	net_per_10K_2019
mean	max	min
base_country_wb_region			
East Asia & Pacific	0.198827	21.57	-9.88
Europe & Central Asia	0.208974	87.71	-21.34
Latin America & Caribbean	-0.904602	21.15	-31.75
Middle East & North Africa	-0.107655	55.60	-50.33
North America	0.239246	23.20	-0.29
South Asia	-0.514577	13.72	-24.89
Sub-Saharan Africa	-0.279729	37.11	-21.54
```

5  
```
	net_per_10K_2015	net_per_10K_2019
base_country_wb_income		
High Income	0.02	0.04
Low Income	0.42	-0.05
Lower Middle Income	-0.02	-0.07
Upper Middle Income	-0.03	-0.08
```

6  
```
base_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64 
```

7  
```
net_per_10K_2015	net_per_10K_2016	net_per_10K_2017	net_per_10K_2018	net_per_10K_2019
base_country_wb_income					
High Income	0.505482	0.391379	0.314178	0.379201	0.401470
Low Income	1.876432	0.798270	-0.684865	-0.677784	-0.681459
Lower Middle Income	0.591654	-0.029893	-0.519433	-0.527136	-0.476616
Upper Middle Income	-0.043419	-0.502916	-0.699240	-0.686626	-0.700101
```






In [64]:
import pandas as pd

def migration_stat():
  excel_url = 'https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true'
  df_mig = pd.read_excel(excel_url, sheet_name='Country Migration')
    
  # Set some display options when run
  pd.options.display.max_rows= 20
  
  #1. Describe the dataset
  print(f'Country migration description \n {df_mig.describe()}')

  #2. Show summary information
  print(f'Country migration summary \n {df_mig.info()}')

  #3. Display the mean net per 10K migration in each of the years 2015 to 2019
  net_10k_col = ["net_per_10K_2015", "net_per_10K_2016", "net_per_10K_2017", "net_per_10K_2018", "net_per_10K_2019"]
  print(f'Country migration summary \n {df_mig[net_10k_col].mean()}')

  #4. Display the mean, max and min migration for the year 2019 for each of the regions (base_country_wb_region)
  ds_net_per_10K_2019 = df_mig.groupby("base_country_wb_region")["net_per_10K_2019"]
  disp_stat = ["mean", "max", "min"]
  print(f'2019-Migration summary \n {ds_net_per_10K_2019.agg(disp_stat)}')

  #5. Display the median net migration for the years 2015 and 2019 for the base countries by income level
  ds_net_15_19_country_income = df_mig.groupby("base_country_wb_income")[["net_per_10K_2015", "net_per_10K_2019"]]
  print(f'2015,2019-Migration summary per income \n {ds_net_15_19_country_income.agg(["median"])}')

  #6. Display the number of target countries in each income level
  ds_country_income = df_mig.groupby("base_country_wb_income")["target_country_name"].count()
  print(f'Num of target countries per income \n {ds_country_income}')

  #7. Display the mean net migration for all five years, for each income level
  ds_net_migration_income = df_mig.groupby("base_country_wb_income")[net_10k_col].mean()
  print(f'Mean net migration per income \n {ds_net_migration_income}')


   

migration_stat()

Mean net migration per income 
                         net_per_10K_2015  net_per_10K_2016  net_per_10K_2017  \
base_country_wb_income                                                         
High Income                     0.505482          0.391379          0.314178   
Low Income                      1.876432          0.798270         -0.684865   
Lower Middle Income             0.591654         -0.029893         -0.519433   
Upper Middle Income            -0.043419         -0.502916         -0.699240   

                        net_per_10K_2018  net_per_10K_2019  
base_country_wb_income                                      
High Income                     0.379201          0.401470  
Low Income                     -0.677784         -0.681459  
Lower Middle Income            -0.527136         -0.476616  
Upper Middle Income            -0.686626         -0.700101  


### Exercise 10 - calculating range over a grouped series

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named *Skill Migration*

1.  Display the max for each skill group category of net migration for the year 2017
2.  Assign the max for each skill group category of net migration for the year 2017 to a variable called **max_skill_migration** and print `max_skill_migration`
3.  Create a second variable called **min_skill_migration** and assign to it the min for each skill group category of net migration for the year 2017, print `min_skill_migration`

4.  You now have two series `max_skill_migration` and `min_skill_migration` each of which is a numpy array.  You can perfom calculations on these two series in the same way as you would individual data items.

So, you can calculate the range for each skill category by subtracting the `min_skill_migration` from `max_skill_migration` to get a new series **skill_migration_range**

skill_migration_range = max_skill_migration - min_skill_migration

Try it out.

5.  Now calculate the range for the year 2019
6.  Now calculate the range for countries grouped by base country income level for the year 2015

Test output:  
1 and 2  
```
skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
```

3    
```
skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
```

4  
```
skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64
```

5  
```
skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64
```

6  
```
wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64
```





In [73]:
import pandas as pd

def skill_migration_stat():
  excel_url = 'https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true'
  user_sheet_name = 'Skill Migration'
  df_skill_mig = pd.read_excel(excel_url, sheet_name=user_sheet_name)

  #1. Display the max for each skill group category of net migration for the year 2017
  df_skill_migration_2017 = df_skill_mig.groupby("skill_group_category")["net_per_10K_2017"]
  max_skill_migration = df_skill_migration_2017.max()
  print(f'Max net migration for year 2017 per skill group \n {max_skill_migration}')

  #2. Assign the max for each skill group category of net migration for the year 2017 to 
  #a variable called max_skill_migration and print max_skill_migration
  # Same as 1 above

  #3. Create a second variable called min_skill_migration and assign to it the min for 
  # each skill group category of net migration for the year 2017, print min_skill_migration
  min_skill_migration = df_skill_migration_2017.min()
  print(f'Min net migration for year 2017 per skill group \n {min_skill_migration}')

  #4. You now have two series max_skill_migration and min_skill_migration each of which is a numpy array. 
  #You can perfom calculations on these two series in the same way as you would individual data items.
  #print(f'max_skill_migration describe {df_max_skill_migration.describe()}')


  #5. Now calculate the range for the year 2019
  df_skill_migration_2019 = df_skill_mig.groupby("skill_group_category")["net_per_10K_2019"]

  skill_migration_range_2019 = df_skill_migration_2019.max() - df_skill_migration_2019.min()
  print(f'Skill migration range for year 2019 per skill group \n {skill_migration_range_2019}')

  #6. Calculate the range for countries grouped by base country income level for the year 2015
  df_skill_migration_2015 = df_skill_mig.groupby("wb_income")["net_per_10K_2015"]
  skill_migration_range_2015 = df_skill_migration_2015.max() - df_skill_migration_2015.min()
  print(f'Skill migration range for year 2015 per skill group \n {skill_migration_range_2015}')



skill_migration_stat()

Max net migration for year 2017 per skill group 
 skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
Min net migration for year 2017 per skill group 
 skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
Skill migration range for year 2019 per skill group 
 skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64
Skill migration range for year 2015 per skill group 
 wb_income
High income      